src/layers.c

```diff
@@ -37,55 +56,41 @@ layers_arrange(struct output *output)
 		return;
 	}
 
-	int nr_layers = sizeof(output->layers) / sizeof(output->layers[0]);
+	int nr_layers = sizeof(output->layer_tree) / sizeof(output->layer_tree[0]);
 	for (int i = 0; i < nr_layers; i++) {
-		struct lab_layer_surface *lab_layer_surface;
+		struct wlr_scene_tree *layer = output->layer_tree[i];
 
 		/*
-		 * First we go over the list of surfaces that have
-		 * exclusive_zone set (e.g. statusbars) because we have to
-		 * determine the usable area before processing regular layouts.
+		 * Process exclusive-zone clients before non-exclusive-zone
+		 * clients, so that the latter give way to the former regardless
+		 * of the order in which they were launched.
 		 */
-		wl_list_for_each(lab_layer_surface, &output->layers[i], link) {
-			struct wlr_scene_layer_surface_v1 *scene_layer_surface =
-				lab_layer_surface->scene_layer_surface;
-			if (scene_layer_surface->layer_surface->current.exclusive_zone) {
-				wlr_scene_layer_surface_v1_configure(
-					scene_layer_surface, &full_area, &usable_area);
-			}
-		}
+		arrange_one_layer(output, &full_area, &usable_area, layer, true);
+		arrange_one_layer(output, &full_area, &usable_area, layer, false);
 
-		/* Now we process regular layouts */
-		wl_list_for_each(lab_layer_surface, &output->layers[i], link) {
-			struct wlr_scene_layer_surface_v1 *scene_layer_surface =
-				lab_layer_surface->scene_layer_surface;
-			if (!scene_layer_surface->layer_surface->current.exclusive_zone) {
-				wlr_scene_layer_surface_v1_configure(
-					scene_layer_surface, &full_area, &usable_area);
-			}
-		}
-
-		wlr_scene_node_set_position(&output->layer_tree[i]->node,
-			scene_output->x, scene_output->y);
+		/* Set node position to account for output layout change */
+		wlr_scene_node_set_position(&layer->node, scene_output->x,
+			scene_output->y);
 	}
 
 	memcpy(&output->usable_area, &usable_area, sizeof(struct wlr_box));
 
 	/* Find topmost keyboard interactive layer, if such a layer exists */
-	uint32_t layers_above_shell[] = {
+	uint32_t layers_above_views[] = {
 		ZWLR_LAYER_SHELL_V1_LAYER_OVERLAY,
 		ZWLR_LAYER_SHELL_V1_LAYER_TOP,
 	};
-	size_t nlayers = sizeof(layers_above_shell)
-		/ sizeof(layers_above_shell[0]);
-	struct lab_layer_surface *layer, *topmost = NULL;
+	size_t nlayers = sizeof(layers_above_views) / sizeof(layers_above_views[0]);
+	struct lab_layer_surface *topmost = NULL;
+	struct wlr_scene_node *node;
 	for (size_t i = 0; i < nlayers; ++i) {
-		wl_list_for_each_reverse(layer,
-				&output->layers[layers_above_shell[i]], link) {
-			struct wlr_layer_surface_v1 *layer_surface =
-				layer->scene_layer_surface->layer_surface;
-			if (layer_surface->current.keyboard_interactive) {
-				topmost = layer;
+		struct wlr_scene_tree *tree = output->layer_tree[layers_above_views[i]];
+		/* Iterate in reverse to give most recent node preference */
+		wl_list_for_each_reverse(node, &tree->children, link) {
```

Comment:
This one doesn't check for `!node->data` like the loop above, so whatever route we take we should likely take the same route in both cases.

Reply:
Same.